### PR TITLE
Auth: Don't rotate auth token when requests are cancelled by client

### PR DESCRIPTION
**Pull request by bergquist**
if the client closes the connection we should not
rotate token since the client will never recieve the
new token.

### CODE REVIEW

File: pkg/middleware/middleware.go [hunk 1]
```diff
@@ -229,6 +230,13 @@ func initContextWithToken(authTokenService models.UserTokenService, ctx *models.
 	// Rotate the token just before we write response headers to ensure there is no delay between
 	// the new token being generated and the client receiving it.
 	ctx.Resp.Before(func(w macaron.ResponseWriter) {
+		// if the request is cancelled by the client we
+		// should not try to rotate the token since the
+		// client would not accept any result.
+		if ctx.Context.Req.Context().Err() == context.Canceled {
```
Comment (bergquist) [inline pkg/middleware/middleware.go]: Another approach would be to only rotate on `200`

**Comment by aknuds1** [comment]
I was also wondering vaguely yesterday, when token rotation was mentioned as a possibly reason for clients getting logged out, if they might not receive the new token.

File: pkg/middleware/middleware.go (hunk 1 above)
Comment (aknuds1) [reply pkg/middleware/middleware.go]: Any idea if there's a concrete reason for not rotating on non-200 responses? I.e., might the client not pick up on the new token in those cases?

**Comment by bergquist** [comment]
My current idea for how to validate this is to set up a slow reverse proxy in front of a datasource and query the reverse proxy until problems occur.

**Comment by kylebrandt** [comment]
Could this be abused to keep tokens from rotating indefinitely?

**Comment by bergquist** [comment]
@kylebrandt Good question, but not really. Unless the token is rotated the user will be denied access eventually. https://github.com/grafana/grafana/blob/master/pkg/services/auth/auth_token.go#L114

File: pkg/middleware/middleware_test.go [hunk 2]
```diff
@@ -571,3 +573,88 @@ func middlewareScenario(t *testing.T, desc string, fn scenarioFunc) {
 		fn(sc)
 	})
 }
+
+func TestDontRotateTokensOnCancelledRequests(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	reqContext, _ := initTokenRotationTest(ctx)
```
Comment (aknuds1) [inline pkg/middleware/middleware_test.go]: I'd prefer if we don't ignore the error here. I can update the branch with an error check?

File: pkg/middleware/middleware_test.go [hunk 3]
```diff
@@ -571,3 +573,88 @@ func middlewareScenario(t *testing.T, desc string, fn scenarioFunc) {
...
+		},
+	}
+
+	token := &models.UserToken{AuthToken: "oldtoken"}
+
+	fn := rotateEndOfRequestFunc(reqContext, uts, token)
+	cancel()
+	fn(reqContext.Resp)
+
+	if tryRotateCallCount > 0 {
```
Comment (aknuds1) [inline pkg/middleware/middleware_test.go]: Let's replace this with `assert.Equal`.

File: pkg/middleware/middleware_test.go [hunk 4]
```diff
@@ -571,3 +573,88 @@ func middlewareScenario(t *testing.T, desc string, fn scenarioFunc) {
...
+	token := &models.UserToken{AuthToken: "oldtoken"}
+
+	rotateEndOfRequestFunc(reqContext, uts, token)(reqContext.Resp)
+
+	foundLoginCookie := false
+	for _, c := range rr.Result().Cookies() {
+		if c.Name == "login_token" {
+			foundLoginCookie = true
+
+			if c.Value == token.AuthToken {
```
Comment (aknuds1) [inline pkg/middleware/middleware_test.go]: Let's use `require.NotEqual` here.

File: pkg/middleware/middleware_test.go [hunk 5]
```diff
@@ -571,3 +573,88 @@ func middlewareScenario(t *testing.T, desc string, fn scenarioFunc) {
...
+				t.Fatal("auth token is still the same")
+			}
+		}
+	}
+
+	if !foundLoginCookie {
```
Comment (aknuds1) [inline pkg/middleware/middleware_test.go]: Let's use `assert.True` here.

**Review by aknuds1 — CHANGES_REQUESTED** [review]
We should use testify assertions in the tests.

File: pkg/middleware/middleware_test.go (hunk 2 above)
Comment (aknuds1) [reply pkg/middleware/middleware_test.go]: Never mind, it's not an error :)

**Review by aknuds1 — APPROVED** [review]
LGTM!

**Review by marefr — APPROVED** [review]
LGTM

File: pkg/middleware/middleware.go (hunk 1 above)
Comment (marefr) [reply pkg/middleware/middleware.go]: Think we want to keep it like this for now. For example if a data source is offline/have problem that shouldn't stop the user from rotating its token.

**Comment by bergquist** [comment]
ref https://github.com/grafana/grafana/issues/15316

**Pull request merged**